ql/src/java/org/apache/hadoop/hive/ql/exec/ddl/DDLTask2.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.exec.ddl;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.hadoop.hive.ql.CompilationOpContext;
+import org.apache.hadoop.hive.ql.DriverContext;
+import org.apache.hadoop.hive.ql.QueryPlan;
+import org.apache.hadoop.hive.ql.QueryState;
+import org.apache.hadoop.hive.ql.exec.Task;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.parse.ExplainConfiguration.AnalyzeState;
+import org.apache.hadoop.hive.ql.plan.api.StageType;
+
+/**
+ * DDLTask implementation.
+**/
+public class DDLTask2 extends Task<DDLWork2> implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private static final Map<Class<? extends DDLDesc>, Class<? extends DDLOperation<?>>> DESC_TO_OPARATION =
+      new HashMap<>();
+  public static void registerOperator(Class<? extends DDLDesc> descClass,
+      Class<? extends DDLOperation<?>> operationClass) {
+    DESC_TO_OPARATION.put(descClass, operationClass);
+  }
+
+  @Override
+  public boolean requireLock() {
+    return this.work != null && this.work.getNeedLock();
+  }
+
+  @Override
+  public void initialize(QueryState queryState, QueryPlan queryPlan, DriverContext ctx,
+      CompilationOpContext opContext) {
+    super.initialize(queryState, queryPlan, ctx, opContext);
+  }
+
+  @Override
+  public int execute(DriverContext driverContext) {
+    if (driverContext.getCtx().getExplainAnalyze() == AnalyzeState.RUNNING) {
+      return 0;
+    }
+
+    try {
+      Hive db = Hive.get(conf);
+      DDLDesc ddlDesc = work.getDDLDesc();
+
+      if (DESC_TO_OPARATION.containsKey(ddlDesc.getClass())) {
+        DDLOperation<?> ddlOperation = DESC_TO_OPARATION.get(ddlDesc.getClass()).newInstance();
```

Comment:
Using a constructor would mean that all the actual Operation classes should declare a constructor just to pass these arguments to the super class, which would lead to a lot of extra code doing nothing specific.

Reply:
if it's not the concreate operation's responsibility then the `init` method should be final.

Reply:
Switched to use constructor. Looks better indeed.